### DISCUSSION TITLE
after Update to XCode 13: error: "Unity-iPhone" requires a provisioning profile.

### DISCUSSION THREAD

**Discussion by IARI** [fastlane/fastlane#19763] [Q&A]
We have upgraded our cloud-mac/build-server
- to macOS Monterey
- uninstalled XCode 12.5, installed XCode 13.2.1

Our Project was set up to use the legacy build system before and now it does not use it anymore.
Our Project is built with Unity 2020.3.25f1 (adding this for the sake of completeness, but this shouldn't really be too important here, Unity just spits out an XCode project after compilation that has to be compiled with XCode)

We have been using fastlane with match to build and sign our app before and it worked fine.
We have three lanes defined with different provisioning profile types. Here is the dev lane:
```
    lane :xcode_build_dev do
        get_api_key
        match(
            type:"development",
            force_for_new_devices: true,
            readonly: false # is_ci
        )
        gym(
          scheme: "Unity-iPhone",
          export_method: "development"
        )
    end
```
The private lane `get_api_key` that basically just calls `app_store_connect_api_key` and shouldn't be of concern.

Everything worked fine before.
Since the upgrade, after match has run successfully. we get:

[15:29:47]: ▸ [35merror: "Unity-iPhone" requires a provisioning profile. Select a provisioning profile in the Signing & Capabilities editor. (in target 'Unity-iPhone' from project 'Unity-iPhone')[0m

Here is the entire build log:
https://pastebin.com/G7e8G6Gd

Why is the provisioning profile from match not used?

**Comment by RobTranquillo** [comment]
We have the same problem here.

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 12.0                                                           |
| Ruby                        | 2.6.8                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.32.0 (Apple Git-132)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | macOS 12.0 (21A5522h)                                          |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 13.2                                                           |
| Swift Version               | 5.5.2                                                          |


### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | de_DE.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |


### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# fastlane/Fastfile

# support local variables
fastlane_require 'dotenv'

before_all do
  Dotenv.overload '.env.secret'
end

# action to run
platform :ios do

  desc "Create .ipa"
    lane :build do
      setup_ci     
      match(
        type: 'development',
        storage_mode: 'git',
        git_url: ENV['MATCH_URL'],
        git_basic_authorization: Base64.strict_encode64("#{ENV['GIT_BASIC_AUTHORIZATION_MAIL']}:#{ENV['MATCH_PERSONAL_ACCESS_TOKEN']}"),
        app_identifier: ENV['IOS_APP_ID']
      )
      update_code_signing_settings(
        use_automatic_signing: true,
        path: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj"
      )
      update_code_signing_settings(
        use_automatic_signing: false,
        team_id: ENV["sigh_#{ENV['IOS_APP_ID']}_appstore_team-id"],
        code_sign_identity: 'iPhone Distribution',
        targets: 'Unity-iPhone',
        path: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj",
        profile_name: ENV["sigh_#{ENV['IOS_APP_ID']}_appstore_profile-name"],
        profile_uuid: ENV["sigh_#{ENV['IOS_APP_ID']}_appstore"]
      )
      build_app(
        project: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj",
        scheme: 'Unity-iPhone',
        xcargs: '-allowProvisioningUpdates',
        export_method: "development"
      )
    end

  end
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# fastlane/Appfile

for_platform :ios do
    app_identifier(ENV['IOS_APP_ID'])

    apple_dev_portal_id(ENV['APPLE_DEVELOPER_EMAIL'])
    itunes_connect_id(ENV['APPLE_CONNECT_EMAIL'])

    team_id(ENV['APPLE_TEAM_ID'])
    itc_team_id(ENV['APPLE_TEAM_ID'])
  end

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.199.0 | ✅ Up-To-Date  |


### Loaded fastlane plugins:

**No plugins Loaded**


<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| did_you_mean                    | 1.3.0        |
| atomos                          | 0.1.3        |
| claide                          | 1.0.3        |
| colored2                        | 3.1.2        |
| nanaimo                         | 0.3.0        |
| rexml                           | 3.2.5        |
| xcodeproj                       | 1.21.0       |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.8.0        |
| terminal-table                  | 1.8.0        |
| plist                           | 3.6.0        |
| public_suffix                   | 4.0.6        |
| addressable                     | 2.8.0        |
| multipart-post                  | 2.0.0        |
| word_wrap                       | 1.0.0        |
| optparse                        | 0.1.1        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| artifactory                     | 3.0.15       |
| babosa                          | 1.0.4        |
| colored                         | 1.2          |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| excon                           | 0.89.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| faraday-rack                    | 1.0.0        |
| ruby2_keywords                  | 0.0.5        |
| faraday                         | 1.8.0        |
| unf_ext                         | 0.0.8        |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| http-cookie                     | 1.0.4        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.2.0        |
| fastimage                       | 2.2.6        |
| gh_inspector                    | 1.1.3        |
| mini_magick                     | 4.11.0       |
| naturally                       | 2.2.1        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| xcpretty-travis-formatter       | 1.0.1        |
| dotenv                          | 2.7.6        |
| bundler                         | 2.3.4        |
| simctl                          | 1.6.8        |
| jwt                             | 2.3.0        |
| uber                            | 0.1.0        |
| declarative                     | 0.0.20       |
| trailblazer-option              | 0.1.2        |
| representable                   | 3.1.1        |
| retriable                       | 3.1.2        |
| mini_mime                       | 1.1.2        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.4        |
| signet                          | 0.16.0       |
| googleauth                      | 1.1.0        |
| httpclient                      | 2.8.3        |
| google-apis-core                | 0.4.1        |
| google-apis-playcustomapp_v1    | 0.6.0        |
| google-apis-androidpublisher_v3 | 0.14.0       |
| google-cloud-env                | 1.5.0        |
| google-cloud-errors             | 1.2.0        |
| google-cloud-core               | 1.6.0        |
| google-apis-iamcredentials_v1   | 0.9.0        |
| google-apis-storage_v1          | 0.10.0       |
| digest-crc                      | 0.6.4        |
| google-cloud-storage            | 1.35.0       |
| emoji_regex                     | 3.2.3        |
| jmespath                        | 1.4.0        |
| aws-partitions                  | 1.541.0      |
| aws-eventstream                 | 1.2.0        |
| aws-sigv4                       | 1.4.0        |
| aws-sdk-core                    | 3.124.0      |
| aws-sdk-kms                     | 1.52.0       |
| aws-sdk-s3                      | 1.109.0      |
| CFPropertyList                  | 3.0.5        |
| json                            | 2.6.1        |
| webrick                         | 1.7.0        |
| rake                            | 13.0.6       |
| forwardable                     | 1.2.0        |
| logger                          | 1.3.0        |
| date                            | 2.0.0        |
| stringio                        | 0.0.2        |
| ipaddr                          | 1.2.2        |
| openssl                         | 2.1.2        |
| zlib                            | 1.0.0        |
| mutex_m                         | 0.1.0        |
| ostruct                         | 0.1.0        |
| strscan                         | 1.0.0        |
| io-console                      | 0.4.7        |
| fileutils                       | 1.1.0        |
| etc                             | 1.0.1        |
| libxml-ruby                     | 3.2.1        |
| psych                           | 3.1.0        |
</details>


*generated on:* **2022-01-05**
</details>

**Reply by IARI** [reply]
I have managed to get builds with _adhoc_ and _appstore_ profiles running, by setting the provisioning profiles in my lane using `update_project_provisioning`.
I have no Idea, why everything worked fine with just using `match` and `gym` without `update_project_provisioning` before, while now with it does not (with XCode 13, new XCode build system, Unity 2020.3.25f1)
but the most important thing is that some ci builds are running again.

I have added this (with the corresponding profile-types and app ID) after `match` and before `gym` inside my lanes

        update_project_provisioning(
            xcodeproj: "Unity-iPhone.xcodeproj",
            profile:ENV["sigh_<MY.APP.ID>_adhoc_profile-path"],
            target_filter: "Unity-iPhone",
        )

To find the exact Environment-Variable names for the profile paths, you can check build logs from earlier failing builds.
If match was executed successfully, it logs out a table showing you the environment variables that it sets.

The reason this seems to work for _adhoc_ and _appstore_ but not _development_ appears to be the certificate.
Match created 2 certificates, one for development and one for _adhoc_ and _appstore_.
Apparently if I do nothing else, the latter one is in use. So development builds fail.

I do know, that `update_project_provisioning` has a `certificate` parameter that expects (jsut like profile)
But from what I understand, match only stores the certificates in the keychain?
(Anyways, as of now I am stuck without development builds)

@RobTranquillo maybe this helps you. in case it does, and you solve the certificate issue let me know ;)

**Reply by RobTranquillo** [reply]
Thanks for your effort, I was able to solve the upper problem without `update_project_provisioning()`. 
I made a very different more basic mistake. Our certificate is "Apple Developer" and the Fastlane default Fastfile I just copied uses "Apple Distribution". But I am too new to this topic to notice this error. And so, my local build runs right now and I hope the github run will be also successful.

**Reply by IARI** [reply]
@RobTranquillo I still could not manage to get development builds running - and its also due to the wrong certificate type being used.
How did you fix the incorrect certificate?

**Reply by RobTranquillo** [reply]
@IARI Hi, I mainly follow the Game-CI/Fastlane documentation. With only some minimal changes to conform to our development certificate. Namely our wildcard certificate what lead to setting IOS_APP_ID to "*". And i do remove AppStore connection because it fails all time. But we don't want to use the AppStore or Testflight.

Here ist our build action

```
      # Game-CI Fastline packaging and signing
      - name: Fix File Permissions and Run Fastlane
        env:
          APPLE_CONNECT_EMAIL: ${{ secrets.APPLE_CONNECT_EMAIL }}
          APPLE_DEVELOPER_EMAIL: ${{ secrets.APPLE_CONNECT_EMAIL }}
          APPLE_TEAM_ID: ${{ secrets.TEAM_ID }}
          APPLE_TEAM_NAME: ${{ secrets.APPLE_TEAM_NAME }}
          MATCH_URL: ${{ secrets.FASTLANE_MATCH_URL }}
          MATCH_PERSONAL_ACCESS_TOKEN: ${{ secrets.FASTLANE_MATCH_PERSONAL_ACCESS_TOKEN }}
          MATCH_GIT_AUTHORIZATION_MAIL: ${{ secrets.FASTLANE_MATCH_GIT_AUTHORIZATION_MAIL }}
          MATCH_PASSWORD: ${{ secrets.FASTLANE_MATCH_PASSWORD }}
          IOS_APP_ID: "*"
          # IOS_APP_ID: de.vrbits.xxxx 
          IOS_BUILD_PATH: ${{ format('{0}/Build', github.workspace) }}
          PROJECT_NAME: xxxxx
        run: |
          find $IOS_BUILD_PATH -type f -name "**.sh" -exec chmod +x {} \;
          bundle install
          bundle exec fastlane ios build
```

and this is our Fastfile:

```  
    desc "Create .ipa"
    lane :build do
      setup_ci
      match(
        type: 'development',
        storage_mode: 'git',
        git_url: ENV['MATCH_URL'],
        git_basic_authorization: Base64.strict_encode64("#{ENV['MATCH_GIT_AUTHORIZATION_MAIL']}:#{ENV['MATCH_PERSONAL_ACCESS_TOKEN']}"),
        app_identifier: ENV['IOS_APP_ID']
      )
      update_code_signing_settings(
        use_automatic_signing: true,
        path: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj"
      )
      update_code_signing_settings(
        use_automatic_signing: false,
        team_id: ENV["sigh_#{ENV['IOS_APP_ID']}_development_team-id"],
        code_sign_identity: 'iPhone Developer',
        targets: 'Unity-iPhone',
        path: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj",
        profile_name: ENV["sigh_#{ENV['IOS_APP_ID']}_development_profile-name"],
        profile_uuid: ENV["sigh_#{ENV['IOS_APP_ID']}_development"]
      )
      build_app(
        project: "#{ENV['IOS_BUILD_PATH']}/iOS/Unity-iPhone.xcodeproj",
        scheme: 'Unity-iPhone',
        xcargs: '-allowProvisioningUpdates',
        export_method: 'development'
      )
    end
```

This is all we use (besides, downloading, building and artifact preservation).
Hope you can find the problem. Get back to me if it doesn't work.

**Reply by IARI** [reply]
I have now managed to also solve the problem for development-type provisioning profiles by adding

        update_code_signing_settings(
            path: "Unity-iPhone.xcodeproj",
            targets: "Unity-iPhone",
            code_sign_identity: "iPhone Developer"
        )

to the corresponding lane.
The reason was, that xcode originally kept trying to use the wrong (distribution type) certificate - the type was set in Unity Player Settings to distribution and consequently also in the XCode project.

Now clearly we cannot have the Unity Player Settings set to one value when the CI is supposed to be able to produce builds with all different kinds of provisioning profile types.
Hence there were two possible solutions:
1. let the C# `BuildCommand` script change the player settings depending on CI parameters
2. use `update_code_signing_settings` in the corresponding lane as depicted above - which is what we did
